i18n/src/main/scala/io/udash/i18n/TranslationKey.scala

```diff
@@ -1,141 +1,134 @@
+  private[i18n] final case class SimpleTranslationKey0(key: String) extends TranslationKey0
+
+  private[i18n] final case class ReducedTranslationKey(key: String, argv: Any*) extends TranslationKey0 {
+    override def apply()(implicit provider: TranslationProvider, lang: Lang): Future[Translated] =
       provider.translate(key, argv: _*)
-    }
+    // default toString puts argv inside "WrappedArray()"
+    override def toString(): String = s"$productPrefix($key,${argv.mkString(",")})"
   }
 
-  private[i18n] final class Untranslatable(key: String) extends TranslationKey0(key) {
+  private [i18n] object ReducedTranslationKey {
+    // note: serialization loses information, as argv is converted to strings
+    implicit val codec: GenCodec[ReducedTranslationKey] = GenCodec.create[ReducedTranslationKey](
+      input => {
+        val list = input.readList()
+        val key = list.nextElement().readSimple().readString()
+        val items = list.iterator(_.readSimple().readString()).toList
+        ReducedTranslationKey(key, items: _*)
+      },
+      (output, value) => {
+        val list = output.writeList()
+        list.writeElement().writeSimple().writeString(value.key)
+        value.argv.foreach(arg => list.writeElement().writeSimple().writeString(arg.toString))
+        list.finish()
+      }
+    )
+  }
+
+  private[i18n] final case class Untranslatable(key: String) extends TranslationKey0 {
     override def apply()(implicit provider: TranslationProvider, lang: Lang): Future[Translated] =
       Future.successful(Translated(key))
   }
 }
 
-object TranslationKey0 {
-  implicit val codec: GenCodec[TranslationKey0] = GenCodec.create[TranslationKey0](
-    (input: Input) => {
-      Try(input.readList()).toOption match {
-        case Some(list) =>
-          val key = list.nextElement().readSimple().readString()
-          val items = list.iterator(_.readSimple().readString()).toList
-          list.skipRemaining()
-          if (items.nonEmpty) new TranslationKey.ReducedTranslationKey(key, items: _*)
-          else TranslationKey.key(key)
-        case None =>
-          TranslationKey.untranslatable(input.readSimple().readString())
-      }
-    },
-    (output: Output, value: TranslationKey0) => {
-      value match {
-        case reduced: TranslationKey.ReducedTranslationKey =>
-          val data = output.writeList()
-          data.writeElement().writeSimple().writeString(value.key)
-          reduced.argv.foreach(item => data.writeElement().writeSimple().writeString(item.toString))
-          data.finish()
-        case untranslatable: TranslationKey.Untranslatable =>
-          output.writeSimple().writeString(untranslatable.key)
-        case std: TranslationKey =>
-          val data = output.writeList()
-          data.writeElement().writeSimple().writeString(std.key)
-          data.finish()
-      }
-    }
-  )
+object TranslationKey0 extends HasGenCodec[TranslationKey0] {
```

Comment:
nit: there should be no reason (historically the way macros worked) for this companion to be at the end of the file